src/devices/Pca9685/Pca9685.cs

```diff
@@ -3,186 +3,354 @@
     {
         /// <summary>
-        /// I2C Device
+        /// I2C address base. Equal to actual address when all selectable address bits are equal to 0.
         /// </summary>
+        public const byte I2cAddressBase = 0x40;
+
         private I2cDevice _device;
 
+        private bool _usingExternalClock;
+        private double _clockFrequency = 25_000_000;
+
         /// <summary>
-        /// Get default clock rate. Set if you are using external clock.
+        /// Get clock frequency (Hz). Only set if you are using external clock.
         /// </summary>
-        public double ClockRate { get; set; } = 25000000;   // 25MHz
+        public double ClockFrequency
+        {
+            get => _clockFrequency;
+            set
+            {
+                if (!_usingExternalClock)
+                    throw new InvalidOperationException("Clock frequency can only be set when using external oscillator.");
+
+                _clockFrequency = value;
+            }
+        }
 
         /// <summary>
         /// Set PWM frequency or get effective value.
         /// </summary>
+        /// <remarks>
+        /// Value of the effective frequency may be different than desired frequency.
+        /// Read the property in order to get the actual value
+        /// </remarks>
         public double PwmFrequency
         {
-            get => _pwmFrequency;
-            set
-            {
-                Prescale = GetPrescale(value);
-            }
+            get => PrescaleToFrequency(_prescale);
+            set => Prescale = FrequencyToPrescale(value);
         }
 
+        private byte _prescale;
+
         /// <summary>
         /// Set PWM frequency using prescale value or get the value.
         /// </summary>
-        public byte Prescale
+        protected byte Prescale
         {
             get => _prescale;
             set
             {
                 var v = value < 3 ? (byte)3 : value;  // min value is 3
-                SetPwmFrequency(v);
+                SetPrescale(v);
                 _prescale = v;
-                _pwmFrequency = GetFreq(v);
             }
         }
 
-        private double _pwmFrequency;
-        private byte _prescale;
-
         /// <summary>
-        /// Initialize PCA9685
+        /// Constructs Pca9685 instance
         /// </summary>
         /// <param name="i2cDevice">The I2C device to be used</param>
-        public Pca9685(I2cDevice i2cDevice)
+        /// <param name="pwmFrequency">Initial PWM frequency on all channels</param>
+        /// <param name="dutyCycleAllChannels">Duty cycle set on all channels after constructing</param>
+        /// <param name="usingExternalClock">
+        /// When true specifies that external clock is used.
+        /// Use <see cref="ClockFrequency"/> to set frequency of an external clock
+        /// </param>
+        public Pca9685(I2cDevice i2cDevice, double pwmFrequency = 100.0, double dutyCycleAllChannels = 0.0, bool usingExternalClock = false)
```

Comment:
PRE_SCALE register has a default value `0x1E` with the PWM frequency is about 196.8876 Hz
Is it good to set `pwmFrequency` to 100.0 ?

Reply:
and so do `dutyCycleAllChannels`

Reply:
Ok, I've just changed the binding to default to -1 for PWM frequency: that will cause prescale to be read from the device and reported back. Same for duty cycle - it won't get changed unless use explicitly changes it